compactor2/src/components/divide_initial/multiple_branches.rs

```diff
@@ -28,14 +28,16 @@ impl DivideInitial for MultipleBranchesDivideInitial {
                 start_level,
                 max_num_files_to_group,
             } => {
-                // Sort start_level files on max_l0_created_at
-                let mut start_level_files = files
+                // To split the start_level files correctly so they can be compacted in the right order,
+                // the files must be sorted on max_l0_created_at if start_level is 0 or min_time otherwise
```

Comment:
Is this elaboration of the comment correct?

```suggestion
                // the files must be sorted on max_l0_created_at if start_level is 0 or min_time otherwise.
                // Since L0s can overlap, they can contain duplicate data, which can only be resolved by
                // using `created_at` time, so the `created_at` time must be used to sort so that it is
                // preserved.  Since L1s & L2s cannot overlap within their own level, they cannot contain
                // duplicate data within their own level, so they do not need to preserve their `created_at`
                // time, so they do not need to be sorted based on `created_at`.  However, sorting by 
                // `min_time` makes it easier to avoid introducing overlaps within their levels.
```

If what I wrote above is correct, I think this means `created_at` is not preserved in L1 & L2 files.

(if this is correct, I have a followup question, but I didn't want to run too far if I'm already misunderstanding...)

Reply:
There are 2 fields for a file:`created_at` and `max_l0_created_at`.
- `create_at` is the actual created time of a file.
- `max_l0_created_at` is the max `created_at` of all L0 files that are compacted into L0 or L1 or L2.

This  means the `created_at` of L1 and L2 do not mean much and is not used in either compactor or Querier where deduplication happen. We use `max_l0_created_at` instead.

We use `max_l0_created_at` to sort L0 so they can be deduplicated correctly as in your comment but `created_at` must be replaced with `max_l0_created_at`as follows  because L0s are also compacted into L0s:

// Since L0s can overlap, they can contain duplicate data, which can only be resolved by
 // using `max_lo_created_at` time, so the `max_l0_created_at` time must be used to sort so that it is
 // preserved

For the second part of your comments which I also replace `created_at` with `max_l0_created_at`:

// Since L1s & L2s cannot overlap within their own level, they cannot contain
                // duplicate data within their own level, so they do not need to preserve their `max_l0_created_at`
                // time, so they do not need to be sorted based on `max_l0_created_at`.  However, sorting by 
                // `min_time` makes it easier to avoid introducing overlaps within their levels.

I think they are correct but we want to replace the word `easier` with something stronger because it is not easier but `a must` to sort on their time to avoid overlaps.

Reply:
And I think the `max_l0_created_at` preserves in data creation order in the context of L1 and L2 too but it cannot be used to order L1/l2 with the design that L1s and L2s cannot overlap with own level

Reply:
yes, my references to `created_at` should be `max_l0_created_at`

When first started thinking through my above comment, I was concerned that `min_time` sort order breaks `max_l0_created_at` in the L1 and L2 levels, so we can't use for resolving duplicates between L1 and L2.

Is it true that any time we're compacting L1s or L2s, the group of files being compacted must include all L1s and L2s that overlap?

Reply:
> Is it true that any time we're compacting L1s or L2s, the group of files being compacted must include all L1s and L2s that overlap?

Yes, when we reach here, all the input files are either (all L0s and their overlapped L1s) or (all L1s and their overlapped L2s). The work to group/classified them are done earlier

Reply:
> Yes, when we reach here, all the input files are either (all L0s and their overlapped L1s) or (all L1s and their overlapped L2s). The work to group/classified them are done earlier

great :+1: 

Specifically, I was thinking of scenarios like this:
```
|--L1.1--| |--L1.3--|

    |----L2.2---|
```

Assume the `max_l0_created_at` order is `.1`, `.2`, `.3`.
Assume `L1.1` has at least one duplicate point with `L2.2`
Assume `L1.3` has at least one duplicate point with `L2.2`

If the two L1s are compacted together without the L2, I think we'd forget that `L1.1`'s duplicate point is older/inferior to `L2.2`'s
If all 3 are compacted in a single compaction, `max_l0_created_at` can help us resolve duplicates.

The scenario in this PR made me realize I didn't have a clear understanding of what enables dedupe between L1 and L2 levels.  It doesn't need to be this PR, but I think this is another area to add test cases (which I'm happy to do...)